### ISSUE TITLE
Error: bad argument #1 for 'gsub' (string is not UTF-8)

### ISSUE THREAD

**Issue by LeMoussel** [tatuylonen/wikitextprocessor#301]
**Page:** [ Arsène Lupin](https://fr.wikipedia.org/wiki/Ars%C3%A8ne_Lupin)
**Error:**
```
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'lienWeb') parent ('Modèle:Lien web', {'auteur1': '[[Jacques Derouard]]', 'auteur2': 'François Busnel', 'auteur3': 'Philippe Delaroche', 'titre': "Comment est né le vrai, l'unique Arsène Lupin", 'périodique': "[[L'Express]]", 'date': '1 septembre 2004', 'lire en ligne': 'https://www.lexpress.fr/culture/livre/comment-est-ne-le-vrai-l-unique-arsene-lupin_809439.html', 'consulté le': '9 décembre 2017'}) at ['Arsène Lupin', 'lien web', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'lienWeb') parent ('Modèle:Lien web', {'titre': 'Germanophobie : le retour des revanchards', 'url': 'http://www.slate.fr/story/47141/boches', 'date': '14 décembre 2011', 'site': '[[Slate (magazine)|Slate]]', 'consulté le': '6 octobre 2019'}) at ['Arsène Lupin', 'Lien web', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': 'Maurice Leblanc', 'titre': "Maurice Leblanc nous parle d'Arsène Lupin", 'périodique': 'Gazette de Bayonne, de Biarritz et du Pays basque', 'date': '13 octobre 1932', 'pages': '2', 'lire en ligne': 'https://www.retronews.fr/journal/gazette-de-bayonne-de-biarritz-et-du-pays-basque/13-octobre-1932/343/1242915/2', 'consulté le': '5 décembre 2017'}) at ['Arsène Lupin', 'Citation bloc', 'ARGVAL-2', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': '[[François Forestier]]', 'titre': 'Arsène Lupin est de retour', 'sous-titre': 'qui était vraiment Maurice Leblanc ?', 'périodique': 'BibliObs', 'jour': '11', 'mois': '12', 'année': '2017', 'lire en ligne': 'https://bibliobs.nouvelobs.com/polar/20111215.OBS6821/arsene-lupin-est-de-retour-qui-etait-vraiment-maurice-leblanc.html', 'consulté le': '12 février 2018'}) at ['Arsène Lupin', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': 'Gaston de Pawlowski', 'titre': "L'honnêteté c'est le vol", 'périodique': '[[Comœdia (journal)|Comœdia]]', 'lieu': 'Paris', 'numéro': '394', 'jour': '28', 'mois': '10', 'année': '1908', 'pages': '1', 'lire en ligne': 'https://gallica.bnf.fr/ark:/12148/bpt6k76460477.item', 'consulté le': '13 février 2018'}) at ['Arsène Lupin', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'nom1': 'Robert Belleret', 'titre': 'Au pays de monsieur Lupin', 'périodique': 'Le Monde', 'lieu': 'Paris', 'jour': '22', 'mois': '08', 'année': '2005', 'lire en ligne': 'https://www.lemonde.fr/culture/article/2005/08/22/au-pays-de-monsieur-lupin_681838_3246.html', 'consulté le': '17 février 2018'}) at ['Arsène Lupin', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'périodique': 'Je sais tout', 'lire en ligne': 'https://gallica.bnf.fr/ark:/12148/bpt6k1029808/f720.item', 'jour': '15', 'mois': 'février', 'année': '1907', 'titre': 'Arsène Lupin, Gentleman-Cambrioleur', 'passage': '717'}) at ['Arsène Lupin', '[[link]]', 'article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': 'Pierre Gervasoni', 'titre': 'Avec « Arsène Lupin banquier », les Brigands se refont une santé', 'périodique': '[[lemonde.fr]]', 'date': '27 décembre 2007', 'lire en ligne': 'https://www.lemonde.fr/culture/article/2007/12/27/avec-arsene-lupin-banquier-les-brigands-se-refont-une-sante_993905_3246.html', 'consulté le': '25 mars 2021'}) at ['Arsène Lupin', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'ouvrage') parent ('Modèle:Ouvrage', {'langue': 'en', 'prénom1': 'Jess', 'nom1': 'Nevins', 'préface': 'Alan Moore', 'titre': 'Impossible Territories', 'sous-titre': 'The Unofficial Companion to the League of Extraordinary Gentlemen, The Black Dossier', 'lieu': '[[Austin (Texas)]]', 'éditeur': 'MonkeyBrain', 'mois': 'août', 'année': '2008', 'pages totales': '208', 'isbn': '978-1-932265-24-8'}) at ['Arsène Lupin', 'ouvrage', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Arsène Lupin: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': 'Daniel Aranda', 'titre': 'Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française', 'périodique': "[[Revue d'Histoire littéraire de la France]]", 'volume': '103', 'éditeur': '[[Presses universitaires de France]]', 'date': 'janvier-février 2003', 'isbn': '9782130534655', 'doi': '10.3917/rhlf.031.0111', 'lire en ligne': 'http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm', 'pages': '111-121', 'id': 'aranda2003', 'plume': 'oui', 'issn': '0035-2411'}) at ['Arsène Lupin', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
```

**Test code :**
```python

# https://github.com/tatuylonen/wikitextprocessor/
from wikitextprocessor import Wtp

# https://github.com/tatuylonen/wiktextract
from wiktextract.wxr_context import WiktextractContext
from wiktextract.config import WiktionaryConfig
from wiktextract.page import clean_node

wiki_config = WiktionaryConfig()
wiki_config.dump_file_lang_code = "fr"
wiki_config.capture_language_codes = ["fr", "mul"]
wxr = WiktextractContext(
        wtp=Wtp(
            db_path="fr-wiki-latest.db",
            lang_code="fr",
            project="wikipedia",
        ),
        config=wiki_config,
)

for wiki_page in wxr.wtp.db_conn.execute(
        """
        SELECT title, body
        FROM pages
        WHERE title = 'Arsène Lupin'
        AND redirect_to IS NULL
        """
    ):
        wikipedia_page_title = wiki_page[0]
        wikipedia_page_wikitext = wiki_page[1]
        wxr.wtp.start_page(wikipedia_page_title)

        wiki_nodes = wxr.wtp.parse(text=wikipedia_page_wikitext)
        page_text_content = clean_node(
            wxr=wxr,
            sense_data={},
            wikinode=wiki_nodes,
            collect_links=False,
        )
```

**Comment by kristian-clausal** [comment]
I'm going to download and reconstruct the database for fr.wikipedia and then take a look at this.

**Comment by LeMoussel** [comment]
Reproduction with this wikitext:
```python
    wikitext= """
* {{Article|auteur1=Daniel Aranda|titre=Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française|périodique=[[Revue d'Histoire littéraire de la France]]|volume=103|éditeur=[[Presses universitaires de France]]|date=janvier-février 2003|isbn=9782130534655|doi=10.3917/rhlf.031.0111|lire en ligne=http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm|pages=111-121|id=aranda2003|plume=oui |issn = 0035-2411 }}
    """
    wxr.wtp.start_page("Test page")
    text = wxr.wtp.expand(wikitext)
    print(text)
```

**Comment by kristian-clausal** [comment]
```
$ python testarsene.py                                                              Arsène Lupin: DEBUG: ITALIC not properly closed on the same line at ['Arsène Lupin'] parsing Arsène Lupin/Héritage/Lupinologie
Arsène Lupin: DEBUG: BOLD not properly closed on the same line at ['Arsène Lupin'] parsing Arsène Lupin/Héritage/Lupinologie
Arsène Lupin: DEBUG: ITALIC not properly closed on the same line at ['Arsène Lupin'] parsing Arsène Lupin/Aventures d'Arsène Lupin/Adaptation des aventures d'Arsène Lupin/Pièces de théâtre
```

Unfortunately couldn't reproduce with the given script. I redownloaded the dump and rebuilt the database file (remember you have to delete the .db file, even with `--skip-extraction`).


Testing with the snippet:

```
Arsène Lupin: DEBUG: ITALIC not properly closed on the same line at ['Arsène Lupin']
(venv) kristian@wiktextract-dev:~/Data/frwp:
$ python testarsene.py

* <span class="ouvrage" id="aranda2003">Daniel Aranda, « <cite style="font-style:noruniversitaires de France]], <abbr class="abbr" title="volume">vol.</abbr>&nbsp;103,&lrm; <time class="nowrap" data-sort-value="2003" datetime="2003">janvier-février 2003</time>, <abbr class="abbr" title="pages">p.</abbr>&nbsp;111-121 <small style="line-height:1em;">([[International Standard Book Number|ISBN]]&nbsp;[[Spécial:Ouvrages de référence/9782130534655|<span class="nowrap">9782130534655</span>]], [[International Standard Serial Number|ISSN]]&nbsp;<span class="plainlinks noarchive">[https://portal.issn.org/resource/issn/0035-2411 0035-2411]</span>, [[Digital Object Identifier|DOI]]&nbsp;<span class="plainlinks noarchive nowrap">[https://dx.doi.org/10.3917/rhlf.031.0111 10.3917/rhlf.031.0111]</span>, [http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm lire en ligne])</small><span class="Z3988" title="ctx_ver=Z39.88-2004&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&rft.genre=article&rft.atitle=Maurice+Leblanc+et+la+r%C3%A9surgence+de+la+%C2%AB+s%C3%A9rie+%C2%BB+dans+la+litt%C3%A9rature+romanesque+fran%C3%A7aise&rft.jtitle=Revue+d%27Histoire+litt%C3%A9raire+de+la+France&rft.au=Daniel+Aranda&rft.date=2003&rft.volume=103&rft.pages=111-121&rft.isbn=9782130534655&rft.issn=0035-2411&rft_id=info%3Adoi%2F10.3917%2Frhlf.031.0111&rfr_id=info%3Asid%2Ffr.wikipedia.org%3ATest+page"></span></span>.<span class="nowrap" title="Ouvrage utilisé pour la rédaction de l'article"> [[Fichier:Icon_flatdesign_plume.svg|20px|link=|alt=Ouvrage utilisé pour la rédaction de l'article]]</span>
```

Please try pulling the newest wikitextprocessor.

**Comment by LeMoussel** [comment]
Hmmm... Strange ...

I build the database this way
```shell
cd wiktextract
wiktwords --db-path="../fr-wiki-latest.db" --dump-file-language-code "fr" --skip-extraction ../frwiki-latest-pages-articles.xml.bz2
cd ..
```
And do
```
$ git pull
Updating c9bbad3..f99c758
Fast-forward
 .github/workflows/lint.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
``` 
and I have this error

**Comment by LeMoussel** [comment]
And in `fr-wiki-latest.db` For `Modèle:Article`, 
```sql
SELECT title, namespace_id, redirect_to, need_pre_expand, body, model
FROM pages
WHERE title = 'Modèle:Article'
```
there is this:
```
title           namespace_id    need_pre_expand body                            model
Modèle:Article	10		0	        {{#invoke:Biblio|article}}	wikitext
```

**Comment by kristian-clausal** [comment]
The dump file I got is 6300287989 in size, and the .db file generated from it is 23015948288 in size. The dump file size should be the same (and checking hash is not worth it, it's too variable so size should always change between dumps), the .db file is probably not going to be the same size because the process is not in deterministic order but including that info in case it might be.

**Comment by LeMoussel** [comment]
I have the same size for the dump file (`frwiki-latest-pages-articles.xml.bz2`)
`fr-wiki-latest.db` size is 23 020 396 544

**Comment by kristian-clausal** [comment]
If you have installed wiktextract or wikitextprocessor through `pip`, you might be running those instead of the repo versions.

@xxyzz do you think it could be feasible to add automatic version strings (based on git hashes) into the code that are automatically updated for each commit and which could be printed out ("version xyz of wiktextract, zyx of wikitextprocessor") when running `wiktwords`?

**Comment by LeMoussel** [comment]
It's a good idea to add automatic version strings 

For the installation, I did:
```shell
git clone https://github.com/tatuylonen/wiktextract.git
cd wiktextract
python -m pip install -e .
cd ..
# MàJ: cd wiktextract;git pull;cd ..

git clone --recurse-submodules --shallow-submodules https://github.com/tatuylonen/wikitextprocessor.git
cd wikitextprocessor
python -m pip install -e .
cd ..
# MàJ: cd wikitextprocessor;git pull;cd ..
```  
I will uninstall both packages and reinstall them.

**Comment by LeMoussel** [comment]
`Successfully installed wikitextprocessor-0.4.96 wiktextract-1.99.7`

```shell
pip freeze
```

```
...
-e git+https://github.com/tatuylonen/wikitextprocessor.git@f99c7585a16d8039f84080375f4fcc9f3244f6a5#egg=wikitextprocessor
-e git+https://github.com/tatuylonen/wiktextract.git@7411e9f4a4fa515c0028016f7b5732b0db6ed043#egg=wiktextract
...
```

Hélas, always the mistake. Really weird

**Comment by LeMoussel** [comment]
I got this error in `wikitextprocessor\src\wikitextprocessor\core.py expand function line: 1349`
```python
            # Use the Lua sandbox to execute a Lua macro.  This will initialize
            # the Lua environment and store it in self.lua if it does not
            # already exist (it needs to be re-created for each new page).
            ret = call_lua_sandbox(self, invoke_args, expander, parent, timeout)
```
with
```
invoke_args =('Biblio', 'article')
parent =('Modèle:Article', {'auteur1': 'Daniel Aranda', 'titre': 'Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française', 'périodique': "[[Revue d'Histoire littéraire de la France]]", 'volume': '103', 'éditeur': '[[Presses universitaires de France]]', 'date': 'janvier-février 2003', 'isbn': '9782130534655', 'doi': '10.3917/rhlf.031.0111', 'lire en ligne': 'http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm', 'pages': '111-121', 'id': 'aranda2003', 'plume': 'oui', 'issn': '0035-2411'})
timeout = None
```

**Comment by xxyzz** [comment]
I also don't see any error in "Arsène Lupin" page...

Some suggestions:
- Check your packages are actually installed in `editable` mode: `python -m pip show wikitextprocessor` should display the local git repo path.  wikitextprocessor must be install with `--force-reinstall` if it's already installed as the dependency of wiktextract, this is documented here: https://github.com/tatuylonen/wiktextract?tab=readme-ov-file#install-from-source
- Check how the xml dump file extracted to sqlite db file, it should be created by calling [`process_dump()`](https://github.com/tatuylonen/wikitextprocessor/blob/f99c7585a16d8039f84080375f4fcc9f3244f6a5/src/wikitextprocessor/dumpparser.py#L93):
```python
process_dump(wtp, "xml.bz2_path", {0, 10, 828}, skip_analyze_templates=True)
```
- I don't recommend using `wiktwords` command to process non-Wiktionary dump file, it could overwrite some templates. Don't do this.
- All Lua errors you posted are happened inside `<ref>` tag, and this tag is removed in `clean_node()`. If you still have this error, you could ignore it. The final expanded text should be the same.

`pip freeze` and `git log` could show the commit hash, I think we don't need to show commit in output, it'd be awkward to implement and unnecessary.

**Comment by LeMoussel** [comment]
For the community, some clarifications.

**Install `wiktextract` from local git repo in editable mode.**
```shell
git clone https://github.com/tatuylonen/wiktextract.git
cd wiktextract
python -m pip install --force-reinstall -e .
cd ..
``` 
**Check your `wiktextract` are installed in editable mode**
```shell
python -m pip show wiktextract
```
For example
```
Name: wiktextract
Version: 1.99.7
Summary: Wiktionary dump file parser and multilingual data extractor
Home-page: https://github.com/tatuylonen/wiktextract
Author:
Author-email: Tatu Ylonen <ylo@clausal.com>
License: MIT License
Location: c:\users\appdata\local\programs\python\python310\lib\site-packages
Editable project location: C:\Users\Dev\Python\WikiExtractor\wiktextract
Requires: levenshtein, nltk, pydantic, wikitextprocessor
Required-by
```
**Install `wikitextprocessor` from local git repo in editable mode.**
```shell
git clone --recurse-submodules --shallow-submodules https://github.com/tatuylonen/wikitextprocessor.git
cd wikitextprocessor
python -m pip install --force-reinstall -e .
cd ..
``` 
**Check your `wikitextprocessor` are installed in editable mode**
```shell
python -m pip show wikitextprocessor
```
For example
```
Name: wikitextprocessor
Version: 0.4.96
Summary: Parser and expander for Wikipedia, Wiktionary etc. dump files, with Lua execution support
Home-page: https://github.com/tatuylonen/wikitextprocessor
Author:
Author-email: Tatu Ylonen <ylo@clausal.com>
License: MIT License
Location: c:\users\appdata\local\programs\python\python310\lib\site-packages
Editable project location: C:\Users\Dev\Python\WikiExtractor\wikitextprocessor
Requires: dateparser, lupa, lxml, mediawiki-langcodes, psutil, requests
Required-by: wiktextrac
```

**Comment by LeMoussel** [comment]
**process_dump about namespace id**
```python
process_dump(
    wtp,
   "frwiki-latest-pages-articles.xml.bz2",
    namespace_ids # namespace id, can be found at the start of dump file
)
```

as noted, the namespace ID can be found at the beginning of the  `"frwiki-latest-pages-articles.xml` file.
```
    <namespaces>
      <namespace key="-2" case="first-letter">Média</namespace>
      <namespace key="-1" case="first-letter">Spécial</namespace>
      <namespace key="0" case="first-letter" />
      <namespace key="1" case="first-letter">Discussion</namespace>
      <namespace key="2" case="first-letter">Utilisateur</namespace>
      <namespace key="3" case="first-letter">Discussion utilisateur</namespace>
      <namespace key="4" case="first-letter">Wikipédia</namespace>
      <namespace key="5" case="first-letter">Discussion Wikipédia</namespace>
      <namespace key="6" case="first-letter">Fichier</namespace>
      <namespace key="7" case="first-letter">Discussion fichier</namespace>
      <namespace key="8" case="first-letter">MediaWiki</namespace>
      <namespace key="9" case="first-letter">Discussion MediaWiki</namespace>
      <namespace key="10" case="first-letter">Modèle</namespace>
      <namespace key="11" case="first-letter">Discussion modèle</namespace>
      <namespace key="12" case="first-letter">Aide</namespace>
      <namespace key="13" case="first-letter">Discussion aide</namespace>
      <namespace key="14" case="first-letter">Catégorie</namespace>
      <namespace key="15" case="first-letter">Discussion catégorie</namespace>
      <namespace key="100" case="first-letter">Portail</namespace>
      <namespace key="101" case="first-letter">Discussion Portail</namespace>
      <namespace key="102" case="first-letter">Projet</namespace>
      <namespace key="103" case="first-letter">Discussion Projet</namespace>
      <namespace key="104" case="first-letter">Référence</namespace>
      <namespace key="105" case="first-letter">Discussion Référence</namespace>
      <namespace key="710" case="first-letter">TimedText</namespace>
      <namespace key="711" case="first-letter">TimedText talk</namespace>
      <namespace key="828" case="first-letter">Module</namespace>
      <namespace key="829" case="first-letter">Discussion module</namespace>
      <namespace key="2600" case="first-letter">Sujet</namespace>
    </namespaces>
```
Found in the code, another method
```python
    wtp = Wtp(
        db_path="fr-wiki-latest.db",
        lang_code="fr",
        project="wikipedia",
    )

    wiki_config = WiktionaryConfig()
    wiki_config.dump_file_lang_code = "fr"
    wiki_config.capture_language_codes = ["fr", "mul"]
    wxr = WiktextractContext(wtp, wiki_config)

    namespace_ids = {
        wtp.NAMESPACE_DATA.get(name, {}).get("id", 0)
        for name in wxr.config.save_ns_names
    }
 ```
 which gives the following list `{0, 100, 4, 106, 10, 14, 110, 828}`
 
 what `namespace id` list should i use?

**Comment by xxyzz** [comment]
Use `{0, 10, 828}`, you could add other ids if you want to process them.

**Comment by kristian-clausal** [comment]
AFAIK, that's the pages you want to keep in the database file; so if you don't want to collect "Discussion module" pages, that is left out. Modules, Modéles, main pages.

**Comment by LeMoussel** [comment]
```python
from wikitextprocessor import Wtp
from wikitextprocessor.dumpparser import process_dump

if __name__ == "__main__":
    wtp = Wtp(
        db_path="fr-wiki-latest.db",
        lang_code="fr",
        project="wikipedia",
    )

    namespace_ids = {0,10,828}

    process_dump(
        wtp,
        "frwiki-latest-pages-articles.xml.bz2",
        namespace_ids,
    )

    print(f"# Wikipedia pages collected: {wtp.saved_page_nums()}")

```

**Comment by LeMoussel** [comment]
```
....
2024-08-28 12:37:30,167 INFO:   ... 4680000 raw pages collected
2024-08-28 12:51:30,481 INFO: Analyzing which templates should be expanded before parsing
# Wikipedia pages collected: 4684570
```
`fr-wiki-latest.db` size is 23 015 948 288. Same as your base @kristian-clausal 


Hélas !  After generating a new sqlite database file by calling `process_dump()` I have the same error.
```python
    wikitext= """
{{Article|auteur1=Daniel Aranda|titre=Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française|périodique=[[Revue d'Histoire littéraire de la France]]|volume=103|éditeur=[[Presses universitaires de France]]|date=janvier-février 2003|isbn=9782130534655|doi=10.3917/rhlf.031.0111|lire en ligne=http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm|pages=111-121|id=aranda2003|plume=oui |issn = 0035-2411 }}    """
    
    wiki_config = WiktionaryConfig()
    wiki_config.dump_file_lang_code = "fr"
    wiki_config.capture_language_codes = ["fr", "mul"]
    wxr = WiktextractContext(
        wtp=Wtp(
            db_path="fr-wiki-latest.db",
            lang_code="fr",
            project="wikipedia",
        ),
        config=wiki_config,
    )

   wxr.wtp.start_page("Test page")

    wiki_nodes = wxr.wtp.parse(text=wikitext)
    text = clean_node(
        wxr=wxr,
        sense_data={},
        wikinode=wiki_nodes,
        node_handler_fn=clean_node_handler,
        post_template_fn=post_template_fn,
    )
    print(text)
```
```
Test page: ERROR: LUA error in #invoke('Biblio', 'article') parent ('Modèle:Article', {'auteur1': 'Daniel Aranda', 'titre': 'Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française', 'périodique': "[[Revue d'Histoire littéraire de la France]]", 'volume': '103', 'éditeur': '[[Presses universitaires de France]]', 'date': 'janvier-février 2003', 'isbn': '9782130534655', 'doi': '10.3917/rhlf.031.0111', 'lire en ligne': 'http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm', 'pages': '111-121', 'id': 'aranda2003', 'plume': 'oui', 'issn': '0035-2411'}) at ['Test page', 'Article', '#invoke', '#invoke']
[string "mw_text"]:81: bad argument #1 for 'gsub' (string is not UTF-8)
Articl
```e

**Comment by kristian-clausal** [comment]
I'm sorry, besides double-checking everything, I don't know what could be the cause. I understand your frustration (I myself end up in situations like this a lot, even with wiktextract and wikitextprocessor).

I don't think these have yet to be mentioned in the thread:
- Are you in the correct `venv`?
- Are the packages in the venv the editable repos?
- Are you running the script from the same place you have `fr-wiki-latest.db`?
- Sometimes I have trouble getting venv working with multiplexers (especially when I tried to run both zellij and tmux at the same time).

For me, it usually turns out it's something like this. It's the Anna Karenina principle, all working cloned repos works the same way, but every broken cloned repo is broken in a unique way...

**Comment by LeMoussel** [comment]
Is it possible to set tracebacks?

**Comment by xxyzz** [comment]
The code can't show Lua traceback when error happens inside MedaWiki Lua module due to Lua 5.1 API limitation.

**Comment by LeMoussel** [comment]
I think I found the reason for this error.

**Test Code:**
```python
    wikitext= """
    {{Article
    |titre=Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française
    |périodique=[[Revue d'Histoire littéraire de la France]]
    |date=janvier-février 2003

    |auteur1=Daniel Aranda
    |volume=103
    |éditeur=[[Presses universitaires de France]]
    |isbn=9782130534655
    |doi=10.3917/rhlf.031.0111
    |lire en ligne=http://www.cairn.info/revue-d-histoire-litteraire-de-la-france-2003-1-page-111.htm
    |pages=111-121
    |id=aranda2003
    |plume=oui
    |issn = 0035-2411
    }}
    """

    wxr.wtp.start_page("Test page")
    wiki_nodes = wxr.wtp.parse(text=wikitext)
    text = clean_node(
        wxr=wxr,
        sense_data={},
        wikinode=wiki_nodes,
    )
    print(text)
``` 
produces the error.

If I replace `|date=janvier-février 2003` with `|date=janvier-fevrier 2003`, no more errors with the following text :
`Daniel Aranda, « Maurice Leblanc et la résurgence de la « série » dans la littérature romanesque française », Revue d'Histoire littéraire de la France, Presses universitaires de France, vol. 103, janvier-février 2003, p. 111-121 (ISBN 9782130534655, ISSN 0035-2411, DOI 10.3917/rhlf.031.0111, lire en ligne). [Alt: Ouvrage utilisé pour la rédaction de l'article]`

**Note:**  The date is correctly formatted with the accent `vol. 103, janvier-février 2003, p. 111-121`

**Comment by xxyzz** [comment]
No idea how the hell Windows could screw up the encoding.

I have checked `dumparser.py` code and python, lxml, sqlite docs, still don't have a clue. Damn Windows.

Could you check the text data in sqlite db are in utf8 encoding and your python code files are also in utf8 encoding?

Do you have "lbzcat" or "bzcat" command installed?

Maybe you could do a favor for both of us, try Linux...

**Comment by xxyzz** [comment]
Try this see if helps:
```diff
diff --git a/src/wikitextprocessor/dumpparser.py b/src/wikitextprocessor/dumpparser.py
index 232abd9..16bd340 100644
--- a/src/wikitextprocessor/dumpparser.py
+++ b/src/wikitextprocessor/dumpparser.py
@@ -25,13 +25,16 @@ def decompress_dump_file(
 ) -> Union[subprocess.Popen, bz2.BZ2File]:
     if dump_path.endswith(".bz2"):
         if shutil.which("lbzcat") is None and shutil.which("bzcat") is None:
-            return bz2.open(dump_path, "rb")
+            return bz2.open(dump_path, "rt", encoding="utf-8")
 
         decompress_command = (
             "lbzcat" if shutil.which("lbzcat") is not None else "bzcat"
         )
         p = subprocess.Popen(
-            [decompress_command, dump_path], stdout=subprocess.PIPE
+            [decompress_command, dump_path],
+            stdout=subprocess.PIPE,
+            text=True,
+            encoding="utf-8",
         )
         if p.stdout is not None:
             return p

```

**Comment by kristian-clausal** [comment]
If this turns out to be a Windows-specific encoding issue, thank you for bringing it to our attention. Hopefully xxyzz's fix will be applicable!

**Comment by LeMoussel** [comment]
Python code files are also in utf8 encoding: Yes.
Check the text data in sqlite db are in utf8 encoding: Yes.
I test the encoding with this pragma: [PRAGMA encoding;](https://www.sqlite.org/pragma.html#pragma_encoding). 
This pragma returns the text encoding: `UTF-8` 

The command "bzcat" is installed on my system, I will test xxyzz's fix.

**Comment by LeMoussel** [comment]
With python this uses `bz2.open` 
with `bz2.open(dump_path, "rt", encoding="utf-8")` I have the error:
```
  File "GenerateDB.py", line 14, in <module>
    process_dump(
  File "D:\Developpement\Python\WikiExtractor\wikitextprocessor\src\wikitextprocessor\dumpparser.py", line 122, in process_dump
    parse_dump_xml(wtp, path, namespace_ids)
  File "D:\Developpement\Python\WikiExtractor\wikitextprocessor\src\wikitextprocessor\dumpparser.py", line 54, in parse_dump_xml
    for _, page_element in etree.iterparse(
  File "src\\lxml\\iterparse.pxi", line 208, in lxml.etree.iterparse.__next__
  File "src\\lxml\\iterparse.pxi", line 193, in lxml.etree.iterparse.__next__
  File "src\\lxml\\iterparse.pxi", line 221, in lxml.etree.iterparse._read_more_events
TypeError: reading file objects must return bytes objects

Que je n'ai pas avec `return bz2.open(dump_path, "rb")`
```

**Comment by xxyzz** [comment]
Not check `PRAGMA`'s result, but check the text data's encoding. Python's [sqlite3 doc](https://docs.python.org/3/library/sqlite3.html#how-to-handle-non-utf-8-text-encodings) says non-UTF-8 encoding data could be inserted.

Ahh... I'm out of ideas.

**Comment by LeMoussel** [comment]
I will manually extract `frwiki-latest-pages-articles.xml.bz2` & call `process_dump`  with `skip_extract_dump` set to `True`

**Comment by xxyzz** [comment]
Err... If you set `skip_extract_dump` to `True`, then how .bz2 file is decompressed makes no difference, it's not used.

**Comment by LeMoussel** [comment]
OK, I will do:
```python
parse_dump_xml(wtp, path, namespace_ids)
init_interwiki_map(wtp)

add_default_templates(wtp)
analyze_and_overwrite_pages(
        wtp, overwrite_folders, skip_extract_dump, skip_analyze_templates
)
```

With path set to `frwiki-latest-pages-articles.xml` 

`frwiki-latest-pages-articles.xml` is UTF-8 encoding.
I use the [chardet](https://github.com/chardet/chardet) module.
```shell
$ chardetect  frwiki-latest-pages-articles.xml
frwiki-latest-pages-articles.xml: utf-8 with confidence 0.99
```

**Comment by LeMoussel** [comment]
Same... I'm out of ideas.
Maybe a mistake in the LUA code?

I know a little bit about LUA and I see in some LUA modules the code `
`ustring = "ustring:ustring", & local ustring = require("ustring:ustring")`

I don't know this syntax with the colon (:) with `require`

Can you explain it to me?

**Comment by xxyzz** [comment]
I think you need to confirm the encoding of the text data inserted to sqlite db file first.

**Comment by LeMoussel** [comment]
The encoding of the text data inserted to sqlite is UTF-8.
I suspect there is a problem between Python/lupa -> LUA code. I did this

**wikitextprocessor\src\wikitextprocessor\lua\mw_text.lua**
```lua
function mw_text.trim(s, charset)
   print(s)
   charset = charset or "\r\n\t\f "
   local ret = mw.ustring.gsub(s, "^[" .. charset .. "]*(.-)[" ..
                                  charset .. "]*$", "%1")
   return ret
end
``` 

**Test-utf8.py**
```
    wikitext = """
    {{Lien web
    |titre=Germanophobie : le retour des revanchards
    |url=http://www.slate.fr/story/47141/boches
    |date=14 decembre 2011
    |consulté le= 6 octobre 2019
    }}
    """

    wxr.wtp.start_page("Test page")
    wiki_nodes = wxr.wtp.parse(text=wikitext)
    text = clean_node(
        wxr=wxr,
        sense_data={},
        wikinode=wiki_nodes,
    )
    print(text)    
```
**Output:**
```
14 decembre 2011
14 decembre 2011
2011
decembre
decembre
14
2011
d��cembre
Test page: ERROR: LUA error in #invoke('Biblio', 'lienWeb') parent ('Modèle:Lien web', {'titre': 'Germanophobie : le retour des revanchards', 'url': 'http://www.slate.fr/story/47141/boches', 'date': '14 decembre 2011', 'consulté le': '6 octobre 2019'}) at ['Test page', 'Lien web', '#invoke', '#invoke']
[string "mw_text"]:82: bad argument #1 for 'gsub' (string is not UTF-8)
```

Rhhhooo why `d��cembre`?

**Comment by LeMoussel** [comment]
This is not a problem between Python/lupa -> LUA code !
But rather in the LUA code itself.  But without LUA traceback, it's not easy to debug

**Test code:**
```python
    wxr.wtp.add_page("Modèle:test-template", 10, "{{#invoke:test|functest}}")
    wxr.wtp.add_page(
        "Module:test",
        828,
        """
        local export = {}

        -- Print contents of `tbl`, with indentation.
        -- `indent` sets the initial level of indentation.
        function tprint (tbl, indent)
            if not indent then indent = 0 end
            for k, v in pairs(tbl) do
                formatting = string.rep("  ", indent) .. k .. ": "
                if type(v) == "table" then
                    print(formatting)
                    tprint(v, indent+1)
                elseif type(v) == 'boolean' then
                    print(formatting .. tostring(v))
                else
                    print(formatting .. v)
                end
            end
        end

        function export.functest(frame)
            local args = frame:getParent().args
            tprint(args)
            return tostring(frame.args[0])
        end

        return export
        """,
        )
    wikitext = """
    {{test-template
    |titre=Germanophobie : le retour des revanchards
    |url=http://www.slate.fr/story/47141/boches
    |date=14 decembre 2011
    |consulté le= 6 octobre 2019
    }}
    """
    wxr.wtp.start_page("")
    expanded = wxr.wtp.expand(wikitext)
    print(expanded)
```

**Output:**
```
date: 14 decembre 2011
url: http://www.slate.fr/story/47141/boches
titre: Germanophobie : le retour des revanchards
consulté le: 6 octobre 2019
```

**Comment by xxyzz** [comment]
You could edit the called Lua module pages in sqlite db, add some `print` to find where the error happens.

**Comment by LeMoussel** [comment]
I think I found the reason for this error.
[Module:Biblio/Lien web](https://fr.wikipedia.org/w/index.php?title=Module:Biblio/Lien_web) Line:268
`local dateFormatee = Commun.inscriptionDate( args )`

[Module:Biblio/Commun](https://fr.wikipedia.org/w/index.php?title=Module:Biblio/Commun) Line 488
```python
if date then
   date = date:lower()
```
Paf !  `date:lower()` doesn't handle the encoding properly. `14 décembre 2011 to lower -> 14 d��cembre 2011`
Which then causes the error `Bad argument #1 for 'gsub' (string is not UTF-8)`

`date:lower()` lowercases the date, but ref to my previous question about `ustring:ustring`, I don't know how/what module & function is used with `date:lower()`

**POC**
```python
    wiki_config = WiktionaryConfig()
    wiki_config.dump_file_lang_code = "fr"
    wiki_config.capture_language_codes = ["fr", "mul"]
    wxr = WiktextractContext(
        wtp=Wtp(
            db_path="fr-wiki-latest.db",
            lang_code="fr",
            project="wikipedia",
        ),
        config=wiki_config,
    )

    wxr.wtp.add_page("Modèle:test-template", 10, "{{#invoke:test|functest}}")
    wxr.wtp.add_page(
        "Module:test",
        828,
        """
        local export = {}

        local Lien_web = require( 'Module:Biblio/Lien web' )
        local Commun = require( 'Module:Biblio/Commun' )

        -- Print contents of `tbl`, with indentation.
        -- `indent` sets the initial level of indentation.
        function tprint (tbl, indent)
            if not indent then indent = 0 end
            for k, v in pairs(tbl) do
                formatting = string.rep("  ", indent) .. k .. ": "
                if type(v) == "table" then
                    print(formatting)
                    tprint(v, indent+1)
                elseif type(v) == 'boolean' then
                    print(formatting .. tostring(v))
                else
                    print(formatting .. v)
                end
            end
        end

        function export.functest(frame)
            local args = frame:getParent().args
            tprint(args)

            local date = Commun.validTextArg( args, 'date' )
            if date then
                print(date .. 'to lower -> ' ..date:lower())
            end
        end

        return export
        """,
        )
    wikitext = """
    {{test-template
    |titre=Germanophobie : le retour des revanchards
    |url=http://www.slate.fr/story/47141/boches
    |date=14 décembre 2011
    |consulté le= 6 octobre 2019
    }}
    """
    wxr.wtp.start_page("")
    expanded = wxr.wtp.expand(wikitext)
```

**Comment by xxyzz** [comment]
We use the same `mw.ustring` code from Scribunto, the problem is Lua `string.lower()` can't process unicode string. I think you have to use Linux...

You could try manually change all `date:lower()` to `date:ulower()` in Lua code, maybe this could fix the error. But you will have more similar errors elsewhere.

**Comment by kristian-clausal** [comment]
If Lua unicode manipulation is broken on Windows, that's a problem.

We do a lot of code-manipulation (or did, at least) on Lua code to make it more compatible. Some of it is string manipulation, sometimes we replace lua functions with our own. The problem is always that it's hard to get it all perfectly correct so that nothing breaks; if we replace `string.lower()` with our own in Python, can we guarantee that it returns a correct value?

**Comment by xxyzz** [comment]
That's not a problem. Lua `string` library can't handle unicode, that's why they use `ustring` library. If we replace it we'll be incompatible with MediaWiki.

**Comment by kristian-clausal** [comment]
Then why does `string.lower()` return correct unicode on Linux, on our machines and on fr.wiktionary.org? `date` seems to be a normal `string`.

**Comment by xxyzz** [comment]
IDK why on Linux `string.lower()` behaves like `ustring.lower()`, all Wikimedia servers run on Linux so this problem is not noticeable anywhere.

**Comment by kristian-clausal** [comment]
Do you think it would it be possible to replace `string.lower()` (and other `string` methods) with others? We do replacements for Scribunto-specific libraries, but I don't remember and can't quickly find any replacements for basic Lua standard library stuff. We could put it behind a toggle, like `--use-unicode-strings`.

**Comment by xxyzz** [comment]
I don't recommend wasting more time on this... The hole `string` library is not meant to handle unicode characters, return unicode characters will cause more problems.

**Comment by kristian-clausal** [comment]
Tatu said that getting wikitextprocessor/wiktextract working on Windows is a low priority (also considering that multiprocessing doesn't work on Windows), so I guess I'll be closing this issue, unfortunately.

**Comment by xxyzz** [comment]
Multiprocessing works on Windows now. The problem of this issue is that French Wikipedia Lua module uses a wrong API.

**Comment by kristian-clausal** [comment]
@LeMoussel if you want to try to figure something out regarding this specifically for this error, take a look at the code in `src/wikitextprocessor/luaexec.py` and `src/wikitextprocessor/lua/_sandbox_phase1.lua` and `_sandbox_phase2.lua`. It might be possible to make a function wrapper around `gsub` (`__orig_gsub` being called inside a wrapper function) so that the string is converted back to utf-8 before being fed to the original `gsub`. There's a bunch of these kinds of functions and wrappers (I think, might have been removed at some point) that you can make in python and pass into the lua-code. This is just a stop-gap measure, however, and it would be pretty messy.

**Comment by xxyzz** [comment]
That's not a good advice... I don't think you can convert them back to utf8 because the string bytes are somehow changed by `string.lower`.

The correct action is fixing the wrong Lua code on Wikipedia.

**Comment by LeMoussel** [comment]
I switched to Linux. No errors.

**Comment by kristian-clausal** [comment]
It is unfortunate that this seems what needs to be done (I'll ask Tatu about windows compatibility later because he thought the multiprocessing problem was still an issue), but good luck with trying out Linux. If you're unfamiliar with Linux, I suggest going with what are the recommended easiest ones: Ubuntu, Linux Mint, etc. because wiktextract and wikitextprocessor definitely don't have as a requirement to be installed on Arch...

**Comment by LeMoussel** [comment]
Thanks. I am using Ubuntu 24.04 LTS.
